review by claude fable 5:

Re-review at 1b0fd6686 — all points from the previous review round are properly addressed.

### What changed
- **In-band signalling removed from the guard**: the assert in `write_chunkindex_to_repo` now tests `entry.flags & ChunkIndex.F_PENDING` instead of comparing `pack_id` against the `UNKNOWN_BYTES32` sentinel — exact, no collision possibility with a legitimate pack ID. The flag is indeed still visible at that point (`hide_system_flags` only masks `M_SYSTEM`/`F_NEW`; flags are stripped only on the serialized batch copy).
- **In-band signalling removed from the test**: it now asserts the persisted entry equals the real in-memory location (`index[H(7)].pack_id == repository.chunks[H(7)].pack_id`), which is also a stronger check than "not the placeholder".
- **Both `UNKNOWN_BYTES32` imports are gone** from `cache.py` and `cache_test.py`.
- **The redundant `flush()` in `Cache.close` was removed** — the flush now lives only in `_maybe_write_chunks_index`, one canonical place. The resulting reordering in `close()` (flush now happens after the files-cache write instead of before) is harmless: the files cache is local and independent of the pack writer, and the flush still precedes the index write on the same path.
- **Test style fixed**: it uses the existing `cache` fixture; the unused `key`/`manifest` parameters are gone.

### Empirical verification (repeated against the new head)
- With the diff applied locally: all 53 tests in `cache_test.py` pass.
- With only the `flush()` line removed: the regression test fails at the new `F_PENDING` assert with a clear message (`AssertionError: chunk 3030…3037 has no pack location yet`), so the guard both catches the bug and diagnoses it well.

### Remaining observations
Nothing blocking. One pre-existing quirk noticed in passing, not introduced by this PR: `Repository.close` asserts the pack writer is empty unconditionally, so an exception that unwinds past a buffered writer could mask the original error with that assert — independent of this change and only reachable if `Cache.close` itself fails mid-way.

**Verdict**: looks ready to merge.

--- a/src/borg/cache.py
+++ b/src/borg/cache.py
@@ -21,7 +21,7 @@ files_cache_logger = create_logger("borg.debug.files_cache")
 from borgstore.store import ItemInfo
 
 from .constants import CACHE_README, FILES_CACHE_MODE_DISABLED, ROBJ_FILE_STREAM, TIME_DIFFERS2_NS
-from .constants import CHUNKINDEX_FRAGMENT_ENTRIES_MIN, CHUNKINDEX_FRAGMENT_ENTRIES_MAX, UNKNOWN_BYTES32
+from .constants import CHUNKINDEX_FRAGMENT_ENTRIES_MIN, CHUNKINDEX_FRAGMENT_ENTRIES_MAX
 from .constants import CHUNKINDEX_SMALL_FRAGMENT_CAP, CHUNKINDEX_MERGE_ATTEMPTS
 from .hashindex import ChunkIndex, ChunkIndexEntry, ChunkIndexEntryFormat
 from .helpers import get_cache_dir
@@ -654,9 +654,9 @@ def write_chunkindex_to_repo(
         batch = ChunkIndex(usable=len(batch_keys) or None)
         for key in batch_keys:
             entry = chunks[key]
-            # a chunk still buffered in the pack writer has UNKNOWN_BYTES32 as its pack_id;
-            # only serialize an entry once its pack is written and its pack_id is real (#9900).
-            assert entry.pack_id != UNKNOWN_BYTES32, f"chunk {bin_to_hex(key)} has no pack location"
+            # a chunk still buffered in the pack writer has F_PENDING set: its pack location
+            # is not resolved yet. only serialize an entry once its pack is written (#9900).
+            assert not (entry.flags & ChunkIndex.F_PENDING), f"chunk {bin_to_hex(key)} has no pack location yet"
             # for now, we don't want to serialize the flags or the size:
             batch[key] = entry._replace(flags=ChunkIndex.F_NONE, size=0)
         new_hash, stored = _store_chunkindex_fragment(repository, batch, stored_hashes, force_write=force_write)
@@ -918,7 +918,7 @@ class ChunksMixin:
         if force or now > self.chunks_index_last_write + self.chunks_index_write_td:
             if self._chunks is not None:
                 # flush the pack writer first, so buffered chunks get their real pack location;
-                # until their pack is written their pack_id is the UNKNOWN_BYTES32 placeholder (#9900).
+                # until their pack is written they are still F_PENDING with no location (#9900).
                 self.repository.flush()
                 write_chunkindex_to_repo(self.repository, self._chunks, clear=clear)
             self.chunks_index_last_write = now
@@ -1011,8 +1011,6 @@ class AdHocWithFilesCache(FilesCacheMixin, ChunksMixin):
     def close(self):
         self.security_manager.save(self.manifest, self.key)
         pi = ProgressIndicatorMessage(msgid="cache.close")
-        if self._chunks is not None:
-            self.repository.flush()
         if self._files is not None:
             pi.output("Saving files cache")
             integrity_data = self._write_files_cache(self._files)

--- a/src/borg/testsuite/cache_test.py
+++ b/src/borg/testsuite/cache_test.py
@@ -20,7 +20,6 @@ from ..cache import (
     repack_chunkindex,
     write_chunkindex_to_repo,
 )
-from ..constants import UNKNOWN_BYTES32
 from ..hashindex import ChunkIndex, ChunkIndexEntry
 from ..crypto.key import AESOCBKey
 from ..helpers import safe_ns
@@ -68,18 +67,20 @@ class TestAdHocWithFilesCache:
         assert cache.add_chunk(H(1), {}, b"5678", stats=Statistics()) == (H(1), 4)
         assert cache.reuse_chunk(H(1), 4, Statistics()) == (H(1), 4)
 
-    def test_periodic_index_write_flushes_pending(self, repository, key, manifest):
+    def test_periodic_index_write_flushes_pending(self, repository, cache):
         # #9900: the periodic index write must flush the pack writer first, so a buffered chunk is
-        # persisted with its real pack location instead of the UNKNOWN_BYTES32 placeholder.
-        cache = AdHocWithFilesCache(manifest)
+        # persisted with its real pack location rather than the unresolved F_PENDING placeholder.
         cache.add_chunk(H(7), {}, b"h1-payload", stats=Statistics())
         assert repository.chunks.is_pending(H(7))  # H(7) is buffered, not yet in a pack
         # force=True runs the same write the 600s timer triggers:
         cache._maybe_write_chunks_index(datetime.now(UTC), force=True)
-        # rebuild the index from the persisted fragments and check H(7)'s stored location:
+        # rebuild the index from the persisted fragments and check H(7)'s stored location matches
+        # the real (flushed) location, not the placeholder:
         index = build_chunkindex_from_repo(repository)
         assert H(7) in index
-        assert index[H(7)].pack_id != UNKNOWN_BYTES32
+        # the flush resolved H(7)'s location, so it is no longer pending/placeholder:
+        assert not repository.chunks.is_pending(H(7))
+        assert index[H(7)].pack_id == repository.chunks[H(7)].pack_id
 
     def test_files_cache(self, cache):
         st = os.stat(".")

